### ISSUE TITLE
Grains on SiloA deactivated because of SiloB shutting down

### ISSUE THREAD

**Issue by Quicksilvr-** [dotnet/orleans#4217]
Hi there.

I've found an issue where app grains on one silo are often being deactivated because another silo shut down. This seems like a serious reliability problem to me, as any grain can essentially be deactivated anywhere in the cluster, at any time, even though their hosting silo is not affected.

Hoping this is not expected behaviour, as any grain that uses timers and does not regularly receive messages could be deactivated in this way.

It's easy to reproduce (see steps at end) and I have been able to repro in v1.5.1 and v1.5.3, using memory and azure table storage (via emulator).

Here are the entries in the log that precede the deactivation:
[2018-03-14 05:57:13.401 GMT    10      INFO    100524  Catalog 10.1.26.61:11111]       Catalog is deactivating 11 activations due to a failure of silo S10.1.26.61:11112:258702790/x177DA9FF, since it is a primary directory partition to these grain ids.
[2018-03-14 05:57:13.407 GMT    14      INFO    103004  VirtualBucketsRingProvider      10.1.26.61:11111]       Removed Server S10.1.26.61:11112:258702790/x177DA9FF. Current view: [S10.1.26.61:11111:258702592 -> <MultiRange: Size=x100000000, %Ring=100.000%>]
[2018-03-14 05:57:13.419 GMT    10      INFO    100541  Catalog 10.1.26.61:11111]       DeactivateActivations: total 11 to shutdown, out of them 11 promptly, 0 later when become idle and 0 are already being destroyed or invalid.
[2018-03-14 05:57:13.425 GMT    14      INFO    103005  VirtualBucketsRingProvider      10.1.26.61:11111]       -NotifyLocalRangeSubscribers about old <MultiRange: Size=x863E5CFB, %Ring=52.439%> new <MultiRange: Size=x100000000, %Ring=100.000%> increased? True
[2018-03-14 05:57:13.442 GMT    10      INFO    100503  Catalog 10.1.26.61:11111]       Starting DestroyActivations #0 of 11 activations
[2018-03-14 05:57:13.447 GMT    14      INFO    100612  MembershipOracle        10.1.26.61:11111]   Will watch (actively ping) 0 silos: []
[2018-03-14 05:57:13.447 GMT    13      INFO    102934  Orleans.Runtime.ReminderService.LocalReminderService    10.1.26.61:11111]    My range changed from <MultiRange: Size=x863E5CFB, %Ring=52.439%> to <MultiRange: Size=x100000000, %Ring=100.000%> increased = True
[2018-03-14 05:57:13.476 GMT     8      ERROR   -1      OrleansTest.PubSubGrain ]       !!!!!!!!!! PubSubGrain on silo S10.1.26.61:11111:258702592 being deactivated because of shutdown of another silo!

The last error above is being logged by me in OnDeactivateAsync(), of the grain that should not be deactivated. It was created on the initial silo and requests delayed deactivation in OnActivateAsync().

I've attached a simple project that makes it easy to repro and debug, by following these steps:
* Open OrleansTest.sln in VS 2017, build and update the debug executable to the OrleansHost.exe in the bin/Debug directory.
* Start the Azure Storage Emulator. Alternatively rename the included OrleansConfiguration_memorystorage.xml and OrleansConfiguration2_memorystorage.xml files to OrleansConfiguration.xml and OrleansConfiguration2.xml, respectively.
* Start Debug session.
* Wait for first silo to start successfully and for one of the messages to be logged, e.g. "Processing message 'some message' in grain ..."
* Open command line in bin/Debug directory.
*## Start the second silo by executing "OrleansHost.exe SecondSilo OrleansConfiguration2.xml".
* Wait for this second silo to start logging "Processing message 'some message' in grain ...". This indicates that a grain on the second silo is also now receiving messages.
* Press Ctrl+C to stop second silo.
* About 60% of the time, the grain sending the messages (which started on the first silo) will be shut down because of the second silo being stopped. The other 40% of the time the grains on the second silo will be activated on the first silo and everything continues working as expected.
* If the incorrect deactivation does not occur, just keep repeating from step *## above and it will happen.

Thanks.

[OrleansDeactivationRepro.zip](https://github.com/dotnet/orleans/files/1809832/OrleansDeactivationRepro.zip)

**Comment by ReubenBond** [comment]
@Quicksilvr- this is something which we would like to address post-2.0

There is discussion and a design which should improve the situation over at #2656. What remains is to reach agreement on the design and to implement it.

Any system which relies on a single machine remaining alive or shutting down gracefully can never be reliable, so you will need to make some changes regardless.

**Comment by Quicksilvr-** [comment]
So as long as more than one silo remains alive then this should not happen? Seems to work fine with 3 instances running and stopping the 3rd.

It would be possible to work around this for specific grains, that do not often get messages (like the timer one I mentioned), by using reminders.

Which leads into another question: Why the restriction on 1 minute as the minimum reminder interval? Would be useful to be able to lower the value for a handful of critical timer based grains...

**Comment by Quicksilvr-** [comment]
Was able to repro, even when using 3 silos and stopping one only one of them.

**Comment by ReubenBond** [comment]
> Which leads into another question: Why the restriction on 1 minute as the minimum reminder interval? Would be useful to be able to lower the value for a handful of critical timer based grains...

Worth asking that in a separate issue - it's worth discussing.

Yes, you'll be able to repro this more easily in small clusters than large clusters because with a small cluster there is a higher chance that a given grain is owned by the directory partition on the destroyed silo.

It's definitely something which we want to fix, hopefully soon

**Comment by Quicksilvr-** [comment]
Thanks for the quick responses. Definitely something that can be managed/worked around in the meantime, for those rare grains that don't receive messages (aside from a bootstrap to activate them perhaps).

I didn't know enough about how the grain directory works, to anticipate this behaviour. Thanks for the link to #2656. The discussion was informative.

**Comment by sergeybykov** [comment]
I'd like to stress what Reuben said above:
>Any system which relies on a single machine remaining alive or shutting down gracefully can never be reliable, so you will need to make some changes regardless.

In my view, in a weird way, the fact that grains get deactivated in this case is good for discovering such assumptions that something will stay alive unless gracefully shut down. It is bad for efficiency because grains get deactivated unnecessarily, and that's why we want to improve this. But an application should recover and stay operational even if some nodes in the cluster fail.

**Comment by Quicksilvr-** [comment]
Agreed. However, I'd bet that I could make a good case for the opposite as well, e.g. "any system where the failure of a single machine can cause cascading 'failures' on other machines in the cluster ..." . ;-)

At the very least, it is counter-intuitive that the failure of a silo could cause grains on "unaffected" silos to be deactivated. It's an Orleans specific implementation detail, which you either know about or you don't. Of course, once you know it, you can take precautions and work around it.

I may have missed it, but can't recall seeing anything in the documentation describing this behaviour/pitfall and I can imagine it could lead to some serious head-scratching should this happen in production, as the effect _appears_ random (affects different sets of grains depending on which silo fails) and is probably masked by whatever caused the silo failure in the first place. Probably something to highlight somewhere, if that hasn't been done already.

I was lucky that I saw this behaviour while testing something else.

**Comment by sergeybykov** [comment]
I agree - this behavior is not intuitive and needs to be documented.

One subtle distinction though. Deactivation of a grain shouldn't be considered a failure. The runtime has a freedom to deactivate and reactivate grains when needed, for example, for migration and rebalancing. We don't do that currently, except for some experiments that were done in [ActOp](https://www.microsoft.com/en-us/research/publication/optimizing-distributed-actor-systems-for-dynamic-interactive-services/?from=http%3A%2F%2Fresearch.microsoft.com%2Fpubs%2F263907%2Feurosys16loca_camera_ready.pdf). But we might in the future.
